File: web/src/lib/utils/album-access.ts
import { getServerErrorMessage } from '$lib/utils/handle-error';
import { AlbumUserRole, getAlbumInfo, type AlbumResponseDto, type UserResponseDto } from '@immich/sdk';

export type AlbumEditAccessResult =
  | { kind: 'allowed'; album: AlbumResponseDto }
  | { kind: 'view_only'; album: AlbumResponseDto }
  | { kind: 'access_denied' }
  | { kind: 'deleted' };

/** First albumUsers entry is always the album owner (SDK contract since v3). */
export const getAlbumOwner = (album: AlbumResponseDto): UserResponseDto | undefined => album.albumUsers[0]?.user;

export const getAlbumOwnerId = (album: AlbumResponseDto): string | undefined => getAlbumOwner(album)?.id;

export const isAlbumEditor = (album: AlbumResponseDto, userId: string): boolean =>
  getAlbumOwnerId(album) === userId ||
  album.albumUsers.some(({ user, role }) => user.id === userId && role === AlbumUserRole.Editor);

export const classifyAlbumAccessError = (error: unknown): 'access_denied' | 'deleted' => {
  const message = getServerErrorMessage(error) ?? '';
  if (/album has been deleted|album not found/i.test(message)) {
    return 'deleted';
  }
  return 'access_denied';
};

export const isAlbumPermissionError = (error: unknown): boolean => {
  const message = getServerErrorMessage(error) ?? '';
  return /album\.read|albumAsset\.create|album has been deleted|album not found/i.test(message);
};

export const checkAlbumEditAccess = async (albumId: string, userId: string): Promise<AlbumEditAccessResult> => {
  try {
    const album = await getAlbumInfo({ id: albumId });
    if (isAlbumEditor(album, userId)) {
      return { kind: 'allowed', album };
    }
    return { kind: 'view_only', album };
  } catch (error) {
    return { kind: classifyAlbumAccessError(error) };
  }
};

export const albumAccessMessageKey = (
  result: Exclude<AlbumEditAccessResult, { kind: 'allowed' }>,
): 'album_permissions_changed_view_only' | 'album_access_denied' | 'album_deleted_by_owner' => {
  switch (result.kind) {
    case 'view_only': {
      return 'album_permissions_changed_view_only';
    }
    case 'access_denied': {
      return 'album_access_denied';
    }
    case 'deleted': {
      return 'album_deleted_by_owner';
    }
  }
};

File: mobile/lib/utils/album_access.dart
import 'dart:convert';

import 'package:immich_mobile/domain/models/album/album.model.dart';
// ignore: import_rule_openapi
import 'package:openapi/api.dart' as openapi;

sealed class AlbumEditAccessResult {
  const AlbumEditAccessResult();
}

final class AlbumEditAccessAllowed extends AlbumEditAccessResult {
  final RemoteAlbum album;
  const AlbumEditAccessAllowed(this.album);
}

final class AlbumEditAccessViewOnly extends AlbumEditAccessResult {
  final RemoteAlbum album;
  const AlbumEditAccessViewOnly(this.album);
}

final class AlbumEditAccessDenied extends AlbumEditAccessResult {
  const AlbumEditAccessDenied();
}

final class AlbumEditAccessDeleted extends AlbumEditAccessResult {
  const AlbumEditAccessDeleted();
}

bool isAlbumEditor(openapi.AlbumResponseDto album, String userId, String ownerId) {
  return ownerId == userId ||
      album.albumUsers.any(
        (albumUser) => albumUser.user.id == userId && albumUser.role == openapi.AlbumUserRole.editor,
      );
}

AlbumEditAccessResult classifyAlbumAccessError(Object error) {
  final message = _serverMessage(error) ?? '';
  if (RegExp(r'album has been deleted|album not found', caseSensitive: false).hasMatch(message)) {
    return const AlbumEditAccessDeleted();
  }
  return const AlbumEditAccessDenied();
}

bool isAlbumPermissionError(Object error) {
  final message = _serverMessage(error) ?? '';
  return RegExp(
    r'album\.read|albumAsset\.create|album has been deleted|album not found',
    caseSensitive: false,
  ).hasMatch(message);
}

String albumAccessMessageKey(AlbumEditAccessResult result) {
  return switch (result) {
    AlbumEditAccessViewOnly() => 'album_permissions_changed_view_only',
    AlbumEditAccessDenied() => 'album_access_denied',
    AlbumEditAccessDeleted() => 'album_deleted_by_owner',
    AlbumEditAccessAllowed() => throw ArgumentError('No message key for allowed access'),
  };
}

String? _serverMessage(Object error) {
  if (error is! openapi.ApiException || error.innerException != null || error.message == null) {
    return null;
  }

  try {
    final body = jsonDecode(error.message!);
    if (body is Map && body['message'] != null) {
      final message = body['message'];
      return message is List ? message.join(', ') : message.toString();
    }
  } catch (_) {}
  return error.message;
}
